src/vm/exceptionhandling.cpp

```diff
@@ -1851,9 +1857,20 @@ CLRUnwindStatus ExceptionTracker::ProcessOSExceptionNotification(
                     ((UINT_PTR)pICF < uCallerSP)) 
                 {
                     pICFForUnwindTarget = pFrame;
+
+                    // When unwinding an exception in ReadyToRun, the JIT_PInvokeEnd helper which unlinks the ICF from 
+                    // the thread will be skipped. This is because unlike jitted code, each pinvoke is wrapped by calls
+                    // to the JIT_PInvokeBegin and JIT_PInvokeEnd helpers, which push and pop the ICF on the thread. The
+                    // ICF is not linked at the method prolog and unlined at the epilog when running R2R code. Since the
+                    // JIT_PInvokeEnd helper will be skipped, we need to unlink the ICF here. If the executing method
+                    // has another pinovoke, it will re-link the ICF again when the JIT_PInvokeBegin helper is called
+
+                    MethodDesc* pPInvokeTargetMD = pFrame->GetFunction();
+                    if (pPInvokeTargetMD != NULL && pPInvokeTargetMD->GetModule()->IsReadyToRun())
```

Comment:
This needs to check whether the *code* being unwound is in R2R image. The code for MethodDesc from R2R image can be JITed for number of reasons.

Reply:
Oh ok, makes sense